[Docs] translate initialize (#687)
open-mmlab/mmengine: docs/en/advanced_tutorials/initialize.md

# Initialization

> Usually, we'll customize our module based on [nn.Module](https://pytorch.org/docs/stable/nn.html?highlight=nn%20module#module-torch.nn.modules), which is implemented by Native PyTorch. Also, [torch.nn.init](https://pytorch.org/docs/stable/nn.init.html?highlight=kaiming#torch.nn.init.kaiming_normal_) could help us initialize the parameters of the model easily. To simplify the process of model construction and initializationm, MMEngine design the [BaseModule](mmengine.model.BaseModule) to help us define and initialize the model from config easily.

Reviewer: ```suggestion
Usually, we'll customize our module based on [nn.Module](https://pytorch.org/docs/stable/nn.html?highlight=nn%20module#module-torch.nn.modules), which is implemented by Native PyTorch. Also, [torch.nn.init](https://pytorch.org/docs/stable/nn.init.html?highlight=kaiming#torch.nn.init.kaiming_normal_) could help us initialize the parameters of the model easily. To simplify the process of model construction and initialization, MMEngine designed the [BaseModule](mmengine.model.BaseModule) to help us define and initialize the model from config easily.
```

## Initialize the model from config

> The core function of `BaseModule` is that it could help us to initialize the model from config. Subclasses inherit from `BaseModule` could define the `init_cfg` in the `__init__` function, and we can choose the method of initialization by configuring `init_cfg`.

Reviewer: ```suggestion
The core function of `BaseModule` is that it could help us to initialize the model from config. Subclasses inherited from `BaseModule` could define the `init_cfg` in the `__init__` function, and we can choose the method of initialization by configuring `init_cfg`.
```

Currently, we support the following initialization methods:

| Initializer                                                     | Registered name | Function                                                                                                                                 |
| :-------------------------------------------------------------- | :-------------: | :--------------------------------------------------------------------------------------------------------------------------------------- |

> | [ConstantInit](../api.html#mmengine.model.ConstantInit)         |    Constant     | Initialize the weight and bias with a constant, commonly used for Convolution                                                            |

Reviewer 2: links failed

| [XavierInit](../api.html#mmengine.model.XavierInit)             |     Xavier      | Initialize the weight by `Xavier` initialization, and initialize the bias with a constant                                                |
| [NormalInit](../api.html#mmengine.model.NormalInit)             |     Normal      | Initialize the weight by normalize distribution, and initialize the bias with a constant                                                 |
| [TruncNormalInit](../api.html#mmengine.model.TruncNormalInit)   |   TruncNormal   | Initialize the weight by truncated normalize distribution, and initialize the bias with a constant，commonly used for Transformer        |

> | [UniformInit](../api.html#mmengine.model.UniformInit)           |     Uniform     | Initialize the weight by union distribution, and initialize the bias with a constant，commonly used for convolution                      |

Reviewer: ```suggestion
| [UniformInit](../api.html#mmengine.model.UniformInit)           |     Uniform     | Initialize the weight by uniform distribution, and initialize the bias with a constant，commonly used for convolution                      |
```

| [KaimingInit](../api.html#mmengine.model.KaimingInit)           |     Kaiming     | Initialize the weight by `Kaiming` initialization, and initialize the bias with a constant. Commonly used for convolution                |
| [Caffe2XavierInit](../api.html#mmengine.model.Caffe2XavierInit) |  Caffe2Xavier   | `Xavier` initialization in Caffe2, and `Kaiming` initialization in PyTorh with `fan_in` and `normal` mode. Commonly used for convolution |
| [PretrainedInit](../api.html#mmengine.model.PretrainedInit)     |   Pretrained    | Initialize the model with the pretrained model                                                                                           |

### Initialize the model with pretrained model

Defining the `ToyNet` as below:

```python
import torch
import torch.nn as nn

from mmengine.model import BaseModule

class ToyNet(BaseModule):

    def __init__(self, init_cfg=None):
        super().__init__(init_cfg)
        self.conv1 = nn.Linear(1, 1)

# Save the checkpoint.
toy_net = ToyNet()
torch.save(toy_net.state_dict(), './pretrained.pth')
pretrained = './pretrained.pth'

toy_net = ToyNet(init_cfg=dict(type='Pretrained', checkpoint=pretrained))
```

> and then we can configure the `init_cfg` to make to load the pretrained model by calling `initi_weights()` after its construction.

Reviewer: ```suggestion
and then we can configure the `init_cfg` to make it load the pretrained model by calling `initi_weights()` after its construction.
```

```python
# Initialize the model with the saved checkpoint.
toy_net.init_weights()
```

```
08/19 16:50:24 - mmengine - INFO - load model from: ./pretrained.pth
08/19 16:50:24 - mmengine - INFO - local loads checkpoint from path: ./pretrained.pth
```

If `init_cfg` is a `dict`, `type` means a kind of initializer registered in `WEIGHT_INITIALIZERS`. The `Pretrained` means `PretrainedInit`, which could help us to load the target checkpoint.
All initializers have the same mapping relationship like `Pretrained` -> `PretrainedInit`, which strips the suffix `Init` of the class name. The `checkpoint` argument of `PretrainedInit` means the path of the checkpoint. It could be a local path or a URL.

### Commonly used initialization methods

Similarly, we could use the `Kaiming` initialization just like `Pretrained` initializer. For example, we could make `init_cfg=dict(type='Kaiming', layer='Conv2d')` to initialize all `Conv2d` module with `Kaiming` initialization.

> Sometimes we need to initialize the model with different initialization methods for different modules. For example, we could initialize the `Conv2d` module with `Kaiming` initialization and initialize the `Linear` module with `Xavier` initialization. We could make `init_cfg=dict(type='Kaiming', layer='Conv2d', override=dict(type='Xavier', layer='Linear'))`:

Reviewer: Is it too early for `override` to occur here?

Author: Stupid me use copilot stupidly 😢 .

conv.weight - torch.Size([1, 1, 1, 1]):
KaimingInit: a=0, mode=fan_out, nonlinearity=relu, distribution =normal, bias=0

08/19 16:50:24 - mmengine - INFO -
conv.bias - torch.Size([1]):
KaimingInit: a=0, mode=fan_out, nonlinearity=relu, distribution =normal, bias=0
```

`layer` could also be a list, each element of which means a type of applied module.

```python
# Apply Kaiming initialization to `Conv2d` and `Linear` module.
toy_net = ToyNet(init_cfg=[dict(type='Kaiming', layer=['Conv2d', 'Linear'])], )
toy_net.init_weights()
```

```
08/19 16:50:24 - mmengine - INFO -
linear.weight - torch.Size([1, 1]):
KaimingInit: a=0, mode=fan_out, nonlinearity=relu, distribution =normal, bias=0

08/19 16:50:24 - mmengine - INFO -
linear.bias - torch.Size([1]):
KaimingInit: a=0, mode=fan_out, nonlinearity=relu, distribution =normal, bias=0

08/19 16:50:24 - mmengine - INFO -
conv.weight - torch.Size([1, 1, 1, 1]):
KaimingInit: a=0, mode=fan_out, nonlinearity=relu, distribution =normal, bias=0

08/19 16:50:24 - mmengine - INFO -
conv.bias - torch.Size([1]):
KaimingInit: a=0, mode=fan_out, nonlinearity=relu, distribution =normal, bias=0
```

### More fine-grained initialization

> Sometimes we need to initialize the same type of module with different type of initialization. For example, we've defined `conv1` and `conv2` submodules, and we want to initialize the `conv1` with `Kaiming` initialization and initialize the `conv2` with `Xavier` initialization. We could configure the init_cfg with `override`:

Reviewer 2: ```suggestion
Sometimes we need to initialize the same type of module with a different types of initialization. For example, we've defined `conv1` and `conv2` submodules, and we want to initialize the `conv1` with `Kaiming` initialization and initialize the `conv2` with `Xavier` initialization. We could configure the init_cfg with `override`:
```

KaimingInit: a=0, mode=fan_out, nonlinearity=relu, distribution =normal, bias=0

08/19 16:50:24 - mmengine - INFO -
conv2.weight - torch.Size([1, 1, 1, 1]):
XavierInit: gain=1, distribution=normal, bias=0

08/19 16:50:24 - mmengine - INFO -
conv2.bias - torch.Size([1]):
KaimingInit: a=0, mode=fan_out, nonlinearity=relu, distribution =normal, bias=0

08/19 16:50:24 - mmengine - INFO -
custom_conv.custom_weight - torch.Size([1, 1, 1, 1]):
Initialized by user-defined `init_weights` in ToyConv

08/19 16:50:24 - mmengine - INFO -
custom_conv.custom_bias - torch.Size([1]):
Initialized by user-defined `init_weights` in ToyConv
```

### Conclusion

**1. Configure `init_cfg` to initialize model**

- Commonly used for the initialization of `Conv2d`, `Linear` and other underlying module. All initialization methods should be managed by `WEIGHT_INITIALIZERS`
- Dynamic initialization controlled by `init_cfg`

**2. Customize `init_weights`**

- Compared to configuring the `init_cfg`, implementing the `init_weights` is simpler and does not require registration. However, it is not as flexible as `init_cfg`, and it is not possible to initialize the module dynamically.

```{note}

> - The priorify of init_weights is higher `init_cfg`

Reviewer: ```suggestion
- The priorify of init_weights is higher than `init_cfg`
```

- Runner will call `init_weights` in Runner.train()
```

### Ininitailize module with function

> As mentioned in prior \[section\](#Customize the initialization method), we could customize our initialization in `init_weights`. To make it more convenient initialize module, MMEngine provides a series of **module initialization functions** to initialize the whole module based on `torch.nn.init`. For example, we want to initialize the weights of the convolutional layer with normal distribution and initialize the bias of the convolutional layer with a constant. The implementation of `torch.nn.init` is as follows:

Reviewer: ```suggestion
As mentioned in prior [section](#Customize the initialization method), we could customize our initialization in `init_weights`. To make it more convenient to initialize modules, MMEngine provides a series of **module initialization functions** to initialize the whole module based on `torch.nn.init`. For example, we want to initialize the weights of the convolutional layer with normal distribution and initialize the bias of the convolutional layer with a constant. The implementation of `torch.nn.init` is as follows:
```

```python
from mmengine.model import kaiming_init, xavier_init

kaiming_init(model)
xavier_init(model)
```

Currently, MMEngine provide the following initialization function:

| initialization function                                               | function                                                                                                                                 |
| :-------------------------------------------------------------------- | :--------------------------------------------------------------------------------------------------------------------------------------- |
| [constant_init](../api.html#mmengine.model.constant_init)             | Initialize the weight and bias with a constant, commonly used for Convolution                                                            |
| [xavier_init](../api.html#mmengine.model.xavier_init)                 | Initialize the weight by `Xavier` initialization, and initialize the bias with a constant                                                |

> | [normal_init](../api.html#mmengine.model.normal_init)                 | Initialize the weight by normalize distribution, and initialize the bias with a constant                                                 |

Reviewer: ```suggestion
| [normal_init](../api.html#mmengine.model.normal_init)                 | Initialize the weight by normal distribution, and initialize the bias with a constant                                                 |
```

> | [trunc_normal_init](../api.html#mmengine.model.trunc_normal_init)     | Initialize the weight by truncated normalize distribution, and initialize the bias with a constant，commonly used for Transformer        |

Reviewer: ```suggestion
| [trunc_normal_init](../api.html#mmengine.model.trunc_normal_init)     | Initialize the weight by truncated normal distribution, and initialize the bias with a constant，commonly used for Transformer        |
```

> | [uniform_init](../api.html#mmengine.model.uniform_init)               | Initialize the weight by union distribution, and initialize the bias with a constant，commonly used for convolution                      |

Reviewer: ```suggestion
| [uniform_init](../api.html#mmengine.model.uniform_init)               | Initialize the weight by uniform distribution, and initialize the bias with a constant，commonly used for convolution                      |
```

| [kaiming_init](../api.html#mmengine.model.kaiming_init)               | Initialize the weight by `Kaiming` initialization, and initialize the bias with a constant. Commonly used for convolution                |
| [caffe2_xavier_init](../api.html#mmengine.model.caffe2_xavier_init)   | `Xavier` initialization in Caffe2, and `Kaiming` initialization in PyTorh with `fan_in` and `normal` mode. Commonly used for convolution |

> | [bias_init_with_prob](../api.html#mmengine.model.bias_init_with_prob) | Initialize the model with the pretrained model                                                                                           |

Reviewer: Explanation seems wrong
